Type family instance expansion chokes when substituting for a DSigT

(Originally reported at goldfirere/singletons/issues/345 .)

@RyanGlScott says at https://github.com/goldfirere/singletons/issues/345#issuecomment-399956160 :

> ```haskell
> {-# LANGUAGE DataKinds #-}
> {-# LANGUAGE PolyKinds #-}
> {-# LANGUAGE TemplateHaskell #-}
> {-# LANGUAGE TypeFamilies #-}
> {-# LANGUAGE TypeOperators #-}
> {-# LANGUAGE UndecidableInstances #-}
> module Bug where
> 
> import GHC.TypeNats
> import Language.Haskell.TH
> import Language.Haskell.TH.Desugar
> 
> type Two = Succ 1
> type family Succ (x :: k) :: k
> type instance Succ (z :: Nat) = z + 1
> 
> $(return [])
> 
> test :: String
> test =
>   $(expandType (DConT ''Two) >>= stringE . show)
> ```
> 
> ```
> $ /opt/ghc/8.4.3/bin/ghci Bug.hs
> GHCi, version 8.4.3: http://www.haskell.org/ghc/  :? for help
> Loaded GHCi configuration from /home/rgscott/.ghci
> [1 of 1] Compiling Bug              ( Bug.hs, interpreted )
> 
> Bug.hs:21:5: error:
>     • Impossible: reification returned a bogus instance
>     • In the untyped splice:
>         $(expandType (DConT ''Two) >>= stringE . show)
>    |
> 21 |   $(expandType (DConT ''Two) >>= stringE . show)
>    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```
> 
> What is going here? When `th-desugar` expands type synonyms, it also makes an effort to expand type family instances as well. So it first expands `Two` to `Succ 1`, and then attempts to expand `Succ 1` using the type family instance for `Succ`.
> 
> However, when attempting to substitute into the RHS of `Succ`'s instance, `th-desugar` tries to match the following two types:
> 
> * `(z :: Nat)`
> * `1`
> 
> However, because of [these lines of code](https://github.com/goldfirere/th-desugar/blob/37ea11115f3b8eb05a7322eb78cc271f4dbfe6b7/Language/Haskell/TH/Desugar/Subst.hs#L122-L125):
> 
> ```haskell
> matchTy ign (DSigT ty _ki) arg = case ign of
>   YesIgnore -> matchTy ign ty arg
>   NoIgnore  -> Nothing
>   -- but we can safely ignore kind signatures on the target
> ```
> 
> Because `expandType` uses `NoIgnore`, it gives up when it encounters the kind signature on `z`. As a result, we get no substitution (i.e., `Nothing`) back in [these lines of code](https://github.com/goldfirere/th-desugar/blob/37ea11115f3b8eb05a7322eb78cc271f4dbfe6b7/Language/Haskell/TH/Desugar/Expand.hs#L130-L134):
> 
> ```haskell
>             case dinsts of
>               [DTySynInstD _n (DTySynEqn lhs rhs)] -> do
>                 subst <-
>                   expectJustM "Impossible: reification returned a bogus instance" $
>                   unionMaybeSubsts $ zipWith (matchTy ign) lhs syn_args
> ```
> 
> Which causes `th-desugar` to panic.
>
> I'm not sure how to proceed here. Perhaps if we receive `Nothing` back here, we should simply give up on expanding type family instances. What are your thoughts, @goldfirere?

@goldfirere comments at https://github.com/goldfirere/singletons/issues/345#issuecomment-400007667 :

> Indeed. No need to panic -- just stop expanding.

--- a/Language/Haskell/TH/Desugar/Expand.hs
+++ b/Language/Haskell/TH/Desugar/Expand.hs
@@ -128,14 +128,13 @@ expand_con ign n args = do
                      qReifyInstances n (map typeToTH syn_args)
             dinsts <- dsDecs insts
             case dinsts of
-              [DTySynInstD _n (DTySynEqn lhs rhs)] -> do
-                subst <-
-                  expectJustM "Impossible: reification returned a bogus instance" $
-                  unionMaybeSubsts $ zipWith (matchTy ign) lhs syn_args
-                ty <- substTy subst rhs
-                ty' <- expand_type ign ty
-                return $ applyDType ty' rest_args
-              _ -> return $ applyDType (DConT n) args
+              [DTySynInstD _n (DTySynEqn lhs rhs)]
+                |  Just subst <-
+                     unionMaybeSubsts $ zipWith (matchTy ign) lhs syn_args
+                -> do ty <- substTy subst rhs
+                      ty' <- expand_type ign ty
+                      return $ applyDType ty' rest_args
+              _ -> give_up
 
 
         DTyConI (DClosedTypeFamilyD (DTypeFamilyHead _n tvbs _frs _ann) eqns) _
@@ -148,7 +147,7 @@ expand_con ign n args = do
               (rhs : _) -> do
                 rhs' <- expand_type ign rhs
                 return $ applyDType rhs' rest_args
-              [] -> return $ applyDType (DConT n) args
+              [] -> give_up
 
           where
              -- returns the substed rhs
@@ -157,7 +156,13 @@ expand_con ign n args = do
               let m_subst = unionMaybeSubsts $ zipWith (matchTy ign) lhs arg_tys
               T.mapM (flip substTy rhs) m_subst
 
-        _ -> return $ applyDType (DConT n) args
+        _ -> give_up
+
+    -- Used when we can't proceed with type family instance expansion any more,
+    -- and must conservatively return the orignal type family applied to its
+    -- arguments.
+    give_up :: q DType
+    give_up = return $ applyDType (DConT n) args
 
     no_tyvars_tyfams :: (DsMonad q, Data a) => a -> q Bool
     no_tyvars_tyfams = everything (liftM2 (&&)) (mkQ (return True) no_tyvar_tyfam)

--- a/Test/Run.hs
+++ b/Test/Run.hs
@@ -252,6 +252,26 @@ test_local_tyfam_expansion =
                    (expandType orig_ty)
        orig_ty `eqTHSplice` exp_ty)
 
+test_stuck_tyfam_expansion :: Bool
+test_stuck_tyfam_expansion =
+  $(do fam_name <- newName "F"
+       x        <- newName "x"
+       k        <- newName "k"
+       let orig_ty = DConT fam_name `DAppT` DConT '() -- F '()
+       exp_ty <- withLocalDeclarations
+                   (decsToTH [ -- type family F (x :: k) :: k
+                               DOpenTypeFamilyD
+                                 (DTypeFamilyHead fam_name
+                                                  [DKindedTV x (DVarT k)]
+                                                  (DKindSig (DVarT k))
+                                                  Nothing)
+                               -- type instance F (x :: ()) = x
+                             , DTySynInstD fam_name
+                                 (DTySynEqn [DSigT (DVarT x) (DConT ''())] (DVarT x))
+                             ])
+                   (expandType orig_ty)
+       orig_ty `eqTHSplice` exp_ty)
+
 test_getDataD_kind_sig :: Bool
 test_getDataD_kind_sig =
 #if __GLASGOW_HASKELL__ >= 800
@@ -424,6 +444,8 @@ main = hspec $ do
     it "works with deriving strategies" $ test_deriving_strategies
 
     it "doesn't expand local type families" $ test_local_tyfam_expansion
+
+    it "doesn't crash on a stuck type family application" $ test_stuck_tyfam_expansion
 
     it "reifies data type return kinds accurately" $ test_getDataD_kind_sig
 

